File: R Programming/tic-tac-toe.R
# This R script will allow you to play a game of Tic-Tac-Toe
# against your computer using the command line.
cat("-----------\nBegin Round\n-----------\n")
Sys.sleep(1)

board <- data.frame(
  c(rep("[ ]", 3)),
  c(rep("[ ]", 3)),
  c(rep("[ ]", 3))
)
colnames(board) <- c(1, 2, 3)

cat("-------------\nCurrent Board\n-------------\n")
board
Sys.sleep(1)

if (interactive()) {
  con <- stdin()
} else {
  con <- "stdin"
}

#function for asking player whether they want to be X or O
get_symbol <- function() {
  cat("X or O? ")
  symbol <- readLines(con = con, n = 1)
  
  while (symbol != "X" & symbol != "O") {
    cat("Invalid input. X or O? ")
    symbol <- readLines(con = con, n = 1)
    Sys.sleep(1)
  }
  return(symbol)
}

#function for asking the player to choose a position for their turn
#and updating the board with their input
update_board_player <- function() {
  cat("Which row? ")
  row_position <- readLines(con = con, n = 1)
  cat("Which column? ")
  col_position <- readLines(con = con, n = 1)
  
  # check if row position or col position entered was empty
  if (is.null(row_position) | is.null(col_position)) {
    valid_position <- FALSE
    # check if position entered is outside the range of the board
    # i.e. not within the 3x3 dataframe
  } else if (is.null(board[row_position, col_position])) {
    valid_position <- FALSE
    # check if position entered is empty on the board
  } else if (board[row_position, col_position] == "[ ]") {
    valid_position <- TRUE
  } else {
    valid_position <- FALSE
  }

  while (valid_position == FALSE) {
    cat("Invalid input. \n")
    cat("Which row? ")
    row_position <- readLines(con = con, n = 1)
    cat("Which column? ")
    col_position <- readLines(con = con, n = 1)
    # check if row position or col position entered was empty
    if (is.null(row_position) | is.null(col_position)) {
      valid_position <- FALSE
      # check if position entered is outside the range of the board
      # i.e. not within the 3x3 dataframe
    } else if (is.null(board[row_position, col_position])) {
      valid_position <- FALSE
      # check if position entered is empty on the board
    } else if (board[row_position, col_position] == "[ ]") {
      valid_position <- TRUE
    } else {
      valid_position <- FALSE
    }
  }
  
  #update board and show it
  board[row_position, col_position] <- player_symbol
  Sys.sleep(1)
  cat("-------------\nCurrent Board\n-------------\n")
  print(board)
  Sys.sleep(1)
  return(board)
}

#function for selecting a position for the computer's turn
update_board_comp <- function() {
  # find which positions on the board are empty
  empty_positions <- which(board == "[ ]", arr.ind = TRUE)
  # choose an random position from the empty positions
  random_position <- sample(nrow(empty_positions), 1)
  row_position <- empty_positions[random_position, ][1]
  col_position <- empty_positions[random_position, ][2]
  #update board and show it
  board[row_position, col_position] <- comp_symbol
  Sys.sleep(1)
  cat("-------------\nCurrent Board\n-------------\n")
  print(board)
  Sys.sleep(1)
  return(board)
}

#This function checks if someone has won by seeing if there are 3 symbols
  #in a row on the board.
check_winner <- function() {
  x_win_vector <- c(rep("X", 3))
  o_win_vector <- c(rep("O", 3))
  empty_positions <- which(board == "[ ]", arr.ind = TRUE)

  end_round <- FALSE

  for (i in 1:nrow(board)) {
    #checking across rows
    if (identical(as.vector(t(board[i, ])), x_win_vector)) {
      print("X wins!")
      end_round <- TRUE
      break
    } else if (identical(as.vector(t(board[i, ])), o_win_vector)) {
      print("O wins!")
      end_round <- TRUE
      break
    #checking across columns
    } else if (identical(board[, i], x_win_vector)) {
      print("X wins!")
      end_round <- TRUE
      break
    } else if (identical(board[, i], o_win_vector)) {
      print("O wins!")
      end_round <- TRUE
      break
    #checking diagonally
    } else if (identical(c(board[1, 1], board[2, 2], board[3, 3]), x_win_vector) |
      identical(c(board[1, 3], board[2, 2], board[3, 1]), x_win_vector)) {
      print("X wins!")
      end_round <- TRUE
      break
    } else if (identical(c(board[1, 1], board[2, 2], board[3, 3]), o_win_vector) |
      identical(c(board[1, 3], board[2, 2], board[3, 1]), o_win_vector)) {
      print("O wins!")
      end_round <- TRUE
      break
    #checking if all positions are filled without a winner
    } else if (end_round == FALSE & length(empty_positions) == 0) {
      print("It's a tie!")
      end_round <- TRUE
      break
    } else {
      end_round <- FALSE
    }
  }
  return(end_round)
}

player_symbol <- get_symbol()

Sys.sleep(1)

end_game <- FALSE

#Player X always goes first, whether they are computer or human.
#Players will alternate taking turns until there is a winner or tie.
if (player_symbol == "X") {
  comp_symbol <- "O"
  while (end_game == FALSE) {
    cat("Player X turn\n")
    board <- update_board_player()
    end_game <- check_winner()
    if (end_game == FALSE) {
      cat("Player O turn\n")
      board <- update_board_comp()
      end_game <- check_winner()
    }
  }
} else {
  comp_symbol <- "X"
  while (end_game == FALSE) {
    cat("Player X turn\n")
    board <- update_board_comp()
    end_game <- check_winner()
    if (end_game == FALSE) {
      cat("Player O turn\n")
      board <- update_board_player()
      end_game <- check_winner()
    }
  }
}

File: Python Programming/tic-tac-toe.py
import random
import logging
import time

# Tic Tac Toe board
board = [[" " for _ in range(3)] for _ in range(3)]


def print_board(board):
    for row in board:
        print("| {} | {} | {} |".format(row[0], row[1], row[2]))


def user_move(symbol):
    while True:
        row = input("Select row for {} (1-3): ".format(symbol))
        col = input("Select column for {} (1-3): ".format(symbol))
        try:
            row = int(row) - 1
            col = int(col) - 1
            if row not in [0, 1, 2] or col not in [0, 1, 2]:
                logging.warning(
                    "Invalid position! Please enter a number from 1 to 3 for both row and column."
                )
            elif board[row][col] != " ":
                logging.warning("This position is already occupied!")
            else:
                board[row][col] = symbol
                break
        except ValueError:
            logging.warning(
                "Invalid input! Please enter a number from 1 to 3 for both row and column."
            )


def computer_move(symbol):
    while True:
        row = random.randint(0, 2)
        col = random.randint(0, 2)
        if board[row][col] == " ":
            board[row][col] = symbol
            break


def check_winner(board, symbol):
    # check rows
    for row in board:
        if row.count(symbol) == 3:
            return True
    # check columns
    for col in range(3):
        if [board[row][col] for row in range(3)].count(symbol) == 3:
            return True
    # check diagonals
    if (
        board[0][0] == board[1][1] == board[2][2] == symbol
        or board[0][2] == board[1][1] == board[2][0] == symbol
    ):
        return True
    return False


def choose_symbol():
    while True:
        symbol = input("Do you want to be X or O? ")
        if symbol not in ["X", "O"]:
            logging.warning("Invalid choice! Please choose either X or O.")
        else:
            return symbol


def play():
    player_symbol = choose_symbol()
    computer_symbol = "O" if player_symbol == "X" else "X"

    print_board(board)

    if player_symbol == "X":
        turn = "user"
    else:
        turn = "computer"

    while True:
        if turn == "user":
            user_move(player_symbol)
            turn = "computer"
        else:
            time.sleep(1)
            print("Computer is making a move...")
            time.sleep(1)
            computer_move(computer_symbol)
            turn = "user"

        print_board(board)
        if check_winner(board, player_symbol):
            print("You Win!")
            break
        elif check_winner(board, computer_symbol):
            print("Computer Wins!")
            break
        elif " " not in [item for sublist in board for item in sublist]:
            print("It's a tie!")
            break


if __name__ == "__main__":
    play()
